platforms/allwinner-d1/core/src/drivers/twi.rs

```diff
@@ -0,0 +1,809 @@
+//! RV Dock, [`TWI2`] is used instead, with `SCL` on pin `PB0` and `SDA` on
+//! pin `PB1`. Separate constructors for the [`TwiI2c0`] type,
+//! [`TwiI2c0::mq_pro`] and [`TwiI2c0::lichee_rv_dock`], are provided to
+//! configure the appropriate pins for each board.
+//!
+//! ## Implementation Notes
+//!
+//! The TWI hardware can be used in one of two modes: "TWI engine" mode, where
+//! individual bytes are written/read from the I<sup>2</sup>C bus in an
+//! interrupt handler, and "TWI driver" mode, where the TWI hardware can operate
+//! at the level of I<sup>2</sup>C register read/writes using a DMA buffer. This
+//! module currently only implements a driver for the "TWI engine" mode, since
+//! it can model all forms of I<sup>2</sup>C operations. In the future, we will
+//! likely want to opportunistically use the offload capabilities of the TWI
+//! driver when the I<sup>2</sup>C transaction has the correct shape for
+//! offloading, but this branch just implements the simpler TWI engine mode.
+//!
+//! The TWI hardware is a bit difficult to use correctly, so implementing this
+//! was a bit of a struggle. In particular, it turns out that the generation of
+//! I<sup>2</sup>C clock pulses occurs when the
+//! [`TWI_CNTR`](d1_pac::twi::TWI_CNTR) register, which controls the TWI, is
+//! written to.
+//!
+//! The driver works by sharing state between a driver task and an ISR, since
+//! the TWI engine mode is interrupt driven. The shared state is "locked" by
+//! disabling TWI interrupts temporarily while the driver task is writing to the
+//! shared state, and resuming interrupts when a write to the shared state
+//! completes. In theory, this driver is also safe for use on multi-core
+//! hardware, since a single daemon task is responsible for all the non-ISR
+//! writes to this shared state, but this doesn't actually matter, because the
+//! D1 is an inherently single-core CPU.
+//!
+//! I believe that the I<sup>2</sup>C controller used in the D1 is from the
+//! Marvell MV64xxx family, although I'm not sure which one in particular. Linux
+//! has a driver for this device, which can be found [here][linux-driver].
+//!
+//! [linux-driver]: https://github.com/torvalds/linux/blob/995b406c7e972fab181a4bb57f3b95e59b8e5bf3/drivers/i2c/busses/i2c-mv64xxx.c
+use core::{
+    cell::UnsafeCell,
+    ops::{Deref, DerefMut},
+    task::{Poll, Waker},
+};
+use d1_pac::{twi, Interrupt, CCU, GPIO, TWI0, TWI1, TWI2, TWI3};
+use kernel::{
+    comms::kchannel::{KChannel, KConsumer},
+    embedded_hal_async::i2c::{ErrorKind, NoAcknowledgeSource},
+    mnemos_alloc::containers::FixedVec,
+    registry,
+    services::i2c::{
+        messages::{OpKind, Transfer},
+        Addr, I2cService, Transaction,
+    },
+    trace, Kernel,
+};
+
+/// A TWI mapped to the Raspberry Pi header's I<sup>2</sup>C0 pins.
+pub struct I2c0 {
+    isr: &'static IsrData,
+    twi: &'static twi::RegisterBlock,
+    /// Which TWI does this TWI Engine use?
```

Comment:
At first I thought this was a todo comment, but it's actually saying "this field owns whatever Interrupt and handler is used for the instance".

But if you can have multiple instances, why is it called I2C0?

Reply:
You can't have multiple instances --- constructing an `I2c0` consumes the `TWI`
from Peripherals, so you can't call it multiple times without passing in
ownership of that peripheral. Which TWI corresponds to the I2C0 pins on the Pi
header depends on the board, which is why we need to have this, to tell the
board-agnostic initialization code which interrupt to unmask.